### PR TITLE
Support root base or home for quickstart #2446

### LINKED ISSUE #2446: AttributeNormalizer does not support "user.home" to be "/"
See https://github.com/eclipse/jetty.project/blob/21365234f8cf036829f7a03e5c7b89a9572444f9/jetty-quickstart/src/main/java/org/eclipse/jetty/quickstart/AttributeNormalizer.java#L89

When https://github.com/eclipse/jetty.project/blob/21365234f8cf036829f7a03e5c7b89a9572444f9/jetty-quickstart/src/main/java/org/eclipse/jetty/quickstart/AttributeNormalizer.java#L228 is "/" I get a 

Caused by: java.lang.IllegalArgumentException: java.net.URISyntaxException: Expected authority at index 7: file://
	at org.eclipse.jetty.quickstart.AttributeNormalizer.toCanonicalURI(AttributeNormalizer.java:93)
	at org.eclipse.jetty.quickstart.AttributeNormalizer.add(AttributeNormalizer.java:208)
	at org.eclipse.jetty.quickstart.AttributeNormalizer.<init>(AttributeNormalizer.java:228)

### PR DESCRIPTION
**Pull request by gregw**
Fix #2446 by only stripping trailing '/' from non-root URIs.  

Signed-off-by: Greg Wilkins <gregw@webtide.com>

### CODE REVIEW

**Comment by gregw** [comment]
@joakime as well as checking this, could you add some more javadoc to AttributeNormalizer.    It's a bit of a strange use-case to start with, so unless one knows the reasons, the class is really hard to read and understand.

**Comment by gregw** [comment]
@joakime can  you review this one.  The build failures are unrelated to the change.

**Comment by joakime** [comment]
Seems that this change is fighting commit 6e1994641b57459808abcadd41f29808ddf2841b which stripped the trailing slash.

The stripping of the trailing slash on that older commit should never have happened.

Argument is ...

The resource `http://machine.com/path/to/resource/` is a valid URI to a directory
The resource `http://machine.com/path/to/resource` is only valid for a file, not a directory.

Same goes for `file://home/joakim/code/` (valid) vs `file://home/joakim/code` (invalid)

As well as `file:/C:/code/` (valid) vs `file:/C:/code` (invalid)
Or `file:/C:/` (valid) vs `file:/C:`

The prior commit should probably just be reverted to fix for issue #2446
The only rationale for the prior commit 6e1994641b57459808abcadd41f29808ddf2841b is for the use of the parameters.

Using `${jetty.home.uri}webapps/root.war` (valid) vs `${jetty.home.uri}/webapps/root.war` (invalid)

**Comment by joakime** [comment]
If we stick with the prior commit 6e1994641b57459808abcadd41f29808ddf2841b and just patch the bad patch.
With a root path, the use of `${jetty.home.uri}/webapps/root.war`  would result in ...

`file://C://webapps/root.war` or `file:////webapps/root.war`  both of which seem awkward.

File: tests/test-quickstart/src/test/java/org/eclipse/jetty/quickstart/AttributeNormalizerTest.java [hunk 1]
```diff
@@ -82,6 +92,16 @@
...
+
+        // ------
+        title = "Root Path Setup";
+        env = new HashMap<>();
+        env.put("jetty.home", "/");
+        env.put("jetty.base", "/");
+        env.put("WAR", asTargetPath(title,"webapps/root"));
+        
+        data.add(new Object[]{arch, title, env});
+
```
Comment (joakime) [inline tests/test-quickstart/src/test/java/org/eclipse/jetty/quickstart/AttributeNormalizerTest.java]: This should be OS neutral.  Consider ..

``` java
        title = "Root Path Setup";
        env = new HashMap<>();
        FileSystem fs = MavenTestingUtils.getTargetPath().getFileSystem();
        Path rootPath = fs.getRootDirectories().iterator().next();
        env.put("jetty.home", rootPath.toString());
        env.put("jetty.base", rootPath.toString());
        env.put("WAR", rootPath.resolve("webapps/root").toString());
```

Comment (joakime) [reply tests/test-quickstart/src/test/java/org/eclipse/jetty/quickstart/AttributeNormalizerTest.java]: Slightly simpler version (and more correct on Windows if running on say `D:\code`)

``` java
        title = "Root Path Setup";
        env = new HashMap<>();
        Path rootPath = MavenTestingUtils.getTargetPath().getRoot();
        env.put("jetty.home", rootPath.toString());
        env.put("jetty.base", rootPath.toString());
        env.put("WAR", rootPath.resolve("webapps/root").toString());
```

**Comment by gregw** [comment]
@joakime can you test on windows again.

Currently, I'm thinking this is not perfect, but it's probably better than what we've currently got.

So the issue we have is that if we generate something like:
```
   ${jetty.base.uri}/resources/my.properties
```
and `jetty.base` is `/` then we end up with 
```
   file:////resources/my.properties
```

A possible solution is to instead generate something like:
```
   ${jetty.base.uri/resources/my.properties}
```
which would expand to the equivalent of
```
   URIUtil.addPaths(properties.get("jetty.base.uri"),"/resources/my.properties")
```
ie it would do the proper additions of path segments, accounting for double /

But I'm reluctant to extend the `${}` syntax without thinking long and hard about it.   I also don't know if we need to do the same kind of thing for Paths?

**Comment by joakime** [comment]
Why does it have to be 

    ${jetty.base.uri}/resources/my.properties

and not the more sane 

    ${jetty.base.uri}resources/my.properties

**Comment by gregw** [comment]
@joakime The problem we have with using `${jetty.base.uri}resources/my.properties` is that it changes the behaviour of `${jetty.base.uri}`.   This is for quickstart and we have generated quickstart files in deployed environments that cannot be regenerated, but have to be able to be run on updated software.   We cannot restrict server updates to require regeneration of deployed quickstarts, so we have to respect existing syntax.   We can add syntax (reluctantly), but we cannot change existing semantics.

**Comment by joakime** [comment]
Can we have the parsing/reading side take up the mantel of supporting existing quickstarts?

Aka ...

``` java

// jetty.base.uri = file:/opt/jetty/mybase/

String existing = "${jetty.base.uri}/resources/my.properties";

String rawParse = util.parse(existing);
// rawParse is "file://opt/jetty/mybase//resources/my.properties"
// Note the double "//" in the raw string

URI asUri = util.toCanonicalURI(rawParse);

// asUri is "file://opt/jetty/mybase/resources/my.properties"
// Note that the canonical URI is fixed (aka normalized) and ready for use.
```

**Comment by gregw** [comment]
@joakime I'm very cautious about taking out `//` as we've had problems when we have done so before as they can be significant sometimes.

I'm inclined to go with the PR as is.... it is not perfect, but the deploy at root use-case will be very rare.  I think what we have is better than before (at least we don't barf with an exception) and I don't think we break any cases that were not already broken.

We need to think carefully before changing anything in the generated files, because once changed it can never be taken back!

**Comment by gregw** [comment]
@joakime what do you think?  More changes or go as is for now?

**Comment by joakime** [comment]
In light of backward compat reasons, and the new fixes for the updated test cases, i'm good with this.

**Pull request merged**